src/lib.rs

```diff
@@ -78,23 +92,39 @@ pin_project! {
 /// Builder for `TlsListener`.
 #[derive(Clone)]
 pub struct Builder {
-    server_config: Arc<ServerConfig>,
+    acceptor: TlsAcceptor,
     max_handshakes: usize,
     handshake_timeout: Duration,
 }
 
+/// Wraps errors from either the listener or the TLS Acceptor
+#[derive(Debug, Error)]
+pub enum Error<A: std::error::Error> {
+    /// An error that arose from the listener ([AsyncAccept::Error])
+    #[error("{0}")]
+    ListenerError(#[source] A),
+    /// An error that occurred during the TLS accept handshake
+    #[cfg(feature = "rustls")]
+    #[error("{0}")]
+    TlsAcceptError(#[source] std::io::Error),
+    /// An error that occurred during the TLS accept handshake
+    #[cfg(feature = "native-tls")]
+    #[error("{0}")]
+    TlsAcceptError(#[source] tokio_native_tls::native_tls::Error),
+}
+
 impl<A: AsyncAccept> TlsListener<A> {
     /// Create a `TlsListener` with default options.
-    pub fn new(server_config: ServerConfig, listener: A) -> Self {
-        builder(server_config).listen(listener)
+    pub fn new(acceptor: TlsAcceptor, listener: A) -> Self {
+        builder(acceptor).listen(listener)
     }
 }
 
 impl<A> TlsListener<A>
 where
     A: AsyncAccept,
-    A::Connection: AsyncRead + AsyncWrite + Unpin,
-    A::Error: Into<Box<dyn Error + Send + Sync>>,
+    A::Connection: AsyncRead + AsyncWrite + Unpin + Sync + Send + 'static,
```

Comment:
The Send + Sync is my mistake, those are not actually needed, so I have removed them. 
I will do some further investigating around the `'static` constraint which is currently needed for the native-tls to compile:
```
error[E0310]: the associated type `<A as AsyncAccept>::Connection` may not live long enough
   --> src\lib.rs:161:29
    |
161 | ...                   Box::pin(async move { tls.accept(conn).await }),
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: consider adding an explicit lifetime bound `<A as AsyncAccept>::Connection: 'static`...
    = note: ...so that the type `impl futures_util::Future<Output = [async output]>` will meet its required lifetime bounds

For more information about this error, try `rustc --explain E0310`.
```